astropy/cosmology/flrw.py

```diff
@@ -141,120 +141,85 @@ def __init__(self, H0, Om0, Ode0, Tcmb0=0.0*u.K, Neff=3.04, m_nu=0.0*u.eV,
-                    self._massivenu_mass = np.full(self._nneutrinos, m_nu.value)
-                    self._nmassivenu = self._nneutrinos
-                    self._nmasslessnu = 0
-            elif len(m_nu) != self._nneutrinos:  # not scalar, check number of masses
+
+            if m_nu.shape not in ((), (self._nneutrinos,)):
                 raise ValueError("unexpected number of neutrino masses — "
                                  f"expected {self._nneutrinos}, got {len(m_nu)}.")
-            # elif m_nu.value.max() == 0:
-            #     pass  # if this is true, min = max = 0 = default case (above)
-            elif np.any(m_nu.value > 0):  # Different masses.
-                massive = np.nonzero(m_nu.value > 0)[0]
-                self._massivenu = True
-                self._nmassivenu = len(massive)
-                self._massivenu_mass = m_nu[massive].value
-                self._nmasslessnu = self._nneutrinos - self._nmassivenu
-
-        # Compute photon density, Tcmb, neutrino parameters
-        # Tcmb0=0 removes both photons and neutrinos, is handled
-        # as a special case for efficiency
-        if self._Tcmb0.value > 0:
-            # Compute photon density from Tcmb
-            self._Ogamma0 = a_B_c2 * self._Tcmb0.value ** 4 / self._critical_density0.value
-
-            # Compute Neutrino temperature
-            # The constant in front is (4/11)^1/3 -- see any
-            #  cosmology book for an explanation -- for example,
-            #  Weinberg 'Cosmology' p 154 eq (3.1.21)
-            self._Tnu0 = 0.7137658555036082 * self._Tcmb0
-
-            # Compute Neutrino Omega and total relativistic component
-            # for massive neutrinos.  We also store a list version,
-            # since that is more efficient to do integrals with (perhaps
-            # surprisingly!  But small python lists are more efficient
-            # than small numpy arrays).
-            if self._massivenu:
-                nu_y = self._massivenu_mass / (kB_evK * self._Tnu0)
-                self._nu_y = nu_y.value
-                self._nu_y_list = self._nu_y.tolist()
-                self._Onu0 = self._Ogamma0 * self.nu_relative_density(0)
-            else:
-                # This case is particularly simple, so do it directly
-                # The 0.2271... is 7/8 (4/11)^(4/3) -- the temperature
-                # bit ^4 (blackbody energy density) times 7/8 for
-                # FD vs. BE statistics.
-                self._Onu0 = 0.22710731766 * self._Neff * self._Ogamma0
-                self._nu_y = self._nu_y_list = None
+            elif np.any(m_nu.value < 0):
+                raise ValueError("invalid (negative) neutrino mass encountered.")
 
-        else:
-            self._Ogamma0 = 0.0
-            self._Tnu0 = 0.0 * u.K
-            self._Onu0 = 0.0
-            self._nu_y = self._nu_y_list = None
+            if m_nu.isscalar:
+                self._m_nu = m_nu = np.full(self._nneutrinos, m_nu.value) << m_nu.unit
```

Comment:
Maybe nicer to do `np.full_like(m_nu, m_nu, shape=self._nneutrinos)` or, if one is sure that this can never be written to, `np.broadcast_to(m_nu, (self._nneutrinos,))`

Reply:
Thanks! I like the latter, but will implement the former. On my long TODO list is to make sure the writeable flag is False for all these parameters. Another layer of immutability.